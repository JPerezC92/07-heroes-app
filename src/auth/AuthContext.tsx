import React, { createContext, useContext, useEffect, useReducer } from "react";
import { AuthAction, authReducer } from "./AuthReducer";

export interface IUser {
  logged: boolean;
  name: string | null;
}

interface IAuthContext extends IUser {
  dispatch: React.Dispatch<AuthAction>;
}

export const initialAuthContextState = {
  logged: false,
  name: null,
} as IAuthContext;

const AuthContext = createContext(initialAuthContextState);

const useAuthState = () => {
  const context = useContext(AuthContext);
  return context;
};

const init = (initialState: IUser) => {
  const userStorageString = localStorage.getItem("user");

  if (userStorageString) return JSON.parse(userStorageString) as IUser;

  return initialState;
};

const AuthContextProvider: React.FC = ({ children }) => {
  const [user, dispatch] = useReducer(
    authReducer,
    initialAuthContextState,
    init
  );

  const value: IAuthContext = { ...user, dispatch };

  useEffect(() => localStorage.setItem("user", JSON.stringify(user)), [user]);

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, useAuthState };
