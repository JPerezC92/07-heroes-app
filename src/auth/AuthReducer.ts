import { IAuthContext } from "./AuthContext";

enum ActionType {
  login = "[auth] login",
  logout = "[auth] logout",
}

export const authReducer = (
  state: IAuthContext,
  action: { type: ActionType; payload: ObjectConstructor }
): IAuthContext => {
  switch (action.type) {
    case ActionType.login:
      return { ...action.payload, logged: true };

    case ActionType.logout:
      return { logged: false };

    default:
      return state;
  }
};
