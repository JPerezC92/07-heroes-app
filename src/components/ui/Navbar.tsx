import { Link, NavLink, useHistory } from "react-router-dom";
import { useAuthState } from "../../auth/AuthContext";
import { AuthActionType } from "../../auth/AuthReducer";

const Navbar = () => {
  const { user, dispatch } = useAuthState();
  const history = useHistory();

  const handleLogout = () => {
    dispatch({ type: AuthActionType.logout });
    history.replace("./login");
  };

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <Link className="navbar-brand" to="/">
        Asociaciones
      </Link>

      <div className="navbar-collapse">
        <div className="navbar-nav">
          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/dc"
          >
            DC
          </NavLink>

          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/search"
          >
            Search
          </NavLink>
        </div>
      </div>

      <div
        className="navbar-collapse collapse w-100 order-3 dual-collapse2"
        style={{
          placeContent: "flex-end",
        }}
      >
        <ul className="navbar-nav ml-auto">
          <span className="nav-item nav-link text-info">{user.name}</span>

          <button className="nav-item nav-link btn" onClick={handleLogout}>
            Logout
          </button>
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
