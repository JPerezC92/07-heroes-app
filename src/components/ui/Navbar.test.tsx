import { mount } from "enzyme";
import { MemoryRouter, Router } from "react-router-dom";
import { useAuthState } from "../../auth/AuthContext";
import { AuthActionType } from "../../auth/AuthReducer";
import { createMemoryHistory } from "history";
import Navbar from "./Navbar";

jest.mock("../../auth/AuthContext", () => {
  return { useAuthState: jest.fn() };
});

describe("Test on <Navbar />", () => {
  const name = "Philip";
  const dispatch = jest.fn();
  const historyMock = createMemoryHistory();
  historyMock.replace = jest.fn();

  (useAuthState as jest.MockedFunction<any>).mockReturnValue({
    user: { logged: true, name },
    dispatch,
  });

  const wrapper = mount(
    <MemoryRouter>
      <Router history={historyMock}>
        <Navbar />
      </Router>
    </MemoryRouter>
  );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".text-info").text().trim()).toBe(name);
  });

  test("should call logout and history.replace", () => {
    expect(wrapper.find("button").exists()).toBeTruthy();
    wrapper.find("button").simulate("click");

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      type: AuthActionType.logout,
    });
    expect(historyMock.replace).toHaveBeenCalledWith("./login");
  });
});
