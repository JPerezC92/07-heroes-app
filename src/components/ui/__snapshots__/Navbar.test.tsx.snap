// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <Navbar /> should render correctly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
    style={
      Object {
        "placeContent": "flex-end",
      }
    }
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-item nav-link text-info"
      >
        Philip
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
